generate_aliases/pkg/generate/generate.go

```diff
@@ -0,0 +1,122 @@
+package generate
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/Dbz/kube-aliases/pkg/types"
+)
+
+func Generate(filePath, targetPath string) error {
+	file, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("error reading file %s with error %s",
+			filePath, err)
+	}
+
+	var aliases types.Aliases
+	err = yaml.Unmarshal(file, &aliases)
+	if err != nil {
+		return fmt.Errorf("error unmarshaling file: %v", err)
+	}
+
+	aliasFile, err := os.Create(targetPath)
+	if err != nil {
+		return fmt.Errorf("Failed to create alias file: %v", err)
+	}
+	defer aliasFile.Close()
+
+	var aliasBuilder strings.Builder
+	aliasNames := map[string]string{}
+	for r := range aliases.Resources {
+
+		// Generate Commands
+		for c := range aliases.CMDs {
+
+			err := GenerateAlias(&aliasBuilder,
+				aliasNames,
+				&types.AliasCMD{
+					PrefixShort:   aliases.CMDs[c].Prefix.Short,
+					ResourceShort: aliases.Resources[r].Short,
+					Short:         aliases.CMDs[c].Short,
+					SuffixShort:   aliases.CMDs[c].Suffix.Short,
+					Prefix:        aliases.CMDs[c].Prefix.CMD,
+					CMD:           aliases.CMDs[c].CMD,
+					Resource:      r,
+					Suffix:        aliases.CMDs[c].Suffix.CMD,
+				})
+			if err != nil {
+				return err
+			}
+
+		}
+
+		for _, v := range aliases.Resources[r].AdditonalCMDs {
+
+			err := GenerateAlias(&aliasBuilder,
+				aliasNames,
+				&types.AliasCMD{
+					PrefixShort:   v.Prefix.Short,
+					ResourceShort: aliases.Resources[r].Short,
+					Short:         v.Short,
+					SuffixShort:   v.Suffix.Short,
+					Prefix:        v.Prefix.CMD,
+					CMD:           v.CMD,
+					Resource:      r,
+					Suffix:        v.Suffix.CMD,
+				})
+
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	_, err = io.WriteString(aliasFile, aliasBuilder.String())
+	if err != nil {
+		return fmt.Errorf("Warning: could not write alias: %v\n", err)
+	}
+
+	return nil
+
+}
+
+// GenerateAlias -- TODO document
+func GenerateAlias(w io.Writer, aliasNames map[string]string, alias *types.AliasCMD) error {
```

Comment:
This works. I feel a little weird about passing in `aliasNames` because it's a leaky abstraction (is that the right term?). I almost feel like it should be part of a larger type that we stick all of the `types.AliasCMD`s into before running this method, and then this method just gets invoked once. 

There might be a more Go way of doing things- I can make a PR to change it.

BTW: I'm a big fan of passing in a writer because that will make the testing way better.

Reply:
Im with you there. I dont particularly like it. But when testing, it was easy to pass a map in. Maybe can make an interface that it acts on. Would probably be better.

Reply:
Refactor PR: https://github.com/Dbz/kube-aliases/pull/26